lib/logstash/plugin_mixins/jdbc/statement_handler.rb

```diff
@@ -29,12 +29,25 @@ class NormalStatementHandler < StatementHandler
     # @param db [Sequel::Database]
     # @param sql_last_value [Integet|DateTime|Time]
     # @yieldparam row [Hash{Symbol=>Object}]
-    def perform_query(db, sql_last_value, jdbc_paging_enabled, jdbc_page_size)
+    def perform_query(db, sql_last_value, jdbc_paging_enabled, jdbc_paging_avoid_count, jdbc_page_size)
       query = build_query(db, sql_last_value)
       if jdbc_paging_enabled
-        query.each_page(jdbc_page_size) do |paged_dataset|
-          paged_dataset.each do |row|
-            yield row
+        if jdbc_paging_avoid_count
+          offset = 0
+          loop do
+            rows_in_page = 0
+            query.with_sql(query.sql, offset: offset, size: jdbc_page_size).each do |row|
```

Comment:
I tried locally and this line expected to be executed one time per page, but `offset` and `size` are not recognized as parameters, so it executes a full scan query and doesn't apply any pagination.
I inserted print statements before and after the query like in:
```ruby
puts "Executing query #{query.sql} offset: #{offset}, size: #{jdbc_page_size}"
query.with_sql(query.sql, offset: offset, size: jdbc_page_size).each do |row|
  yield row
  rows_in_page += 1
end
puts "rows_in_page: #{rows_in_page}"            
```

what I've obtained is:
```
Executing query SELECT * FROM data_log WHERE log_id > 0 offset: 0, size: 7
[2021-11-25T09:38:00,601][INFO ][logstash.inputs.jdbc     ][main][ff8e261c00f8975f017675a993b934c93e346dc2fe1ad66ed9c686980d264c2b] (0.013970s) SELECT * FROM data_log WHERE log_id > 0
rows_in_page: 39
```
while I would have expected the output of multiple lines in the form of:
```
[2021-11-25T09:38:00,601][INFO ][logstash.inputs.jdbc     ][main][ff8e261c00f8975f017675a993b934c93e346dc2fe1ad66ed9c686980d264c2b] (0.013970s) SELECT * FROM data_log WHERE log_id > 0 offset 0 rows 7

[2021-11-25T09:38:00,601+delta][INFO ][logstash.inputs.jdbc     ][main][ff8e261c00f8975f017675a993b934c93e346dc2fe1ad66ed9c686980d264c2b] (0.013970s) SELECT * FROM data_log WHERE log_id > 0 offset 7 rows 7;

[2021-11-25T09:38:00,601+delta][INFO ][logstash.inputs.jdbc     ][main][ff8e261c00f8975f017675a993b934c93e346dc2fe1ad66ed9c686980d264c2b] (0.013970s) SELECT * FROM data_log WHERE log_id > 0 offset 14 rows 7;
```

Reply:
Hi, here's the thing.
When using this paging mode, the part of paging conditions has to be written by the user.
Therefore, the statement should be written as follow:
`
SELECT * FROM data_log WHERE log_id > :sql_last_value LIMIT :size OFFSET :offset
`
I found that it was also missing from my unit test code. I've corrected it.

Reply:
Now I see it. My doubt is that for paged queries where `jdbc_paging_avoid_count` is `false` there is not need to remember to add paging statement in the query, `Sequel` does it. But when the `jdbc_paging_avoid_count` is `true` the user has to remember to add those. 
I think that to provide best experience, the query has to remain the same, so without intervention when switching the `jdbc_paging_avoid_count` on and off. It's to risky to let that in hands of a user, because he can forget to add those, and so inadvertently create maybe killing queries.

Reply:
Hi. Allow me to explain again from the beginning. It takes a little patience to read.

The goal of this PR is to deal with the following situations when `jdbc_paging_enabled` is `true`:

Assume that the database is mysql, and `jdbc_page_size` is `10`.

## 1. simple query
Assume the following statement:
```
SELECT * from test_table
```
The following statements will be created automatically by sequel:
* SELECT count(*) AS count FROM (SELECT * FROM test_table) AS t1 LIMIT 1
* SELECT * FROM (SELECT * FROM test_table) AS t1 LIMIT 10 OFFSET 0
* SELECT * FROM (SELECT * FROM test_table) AS t1 LIMIT 10 OFFSET 10
* subsequent ...

Issues:
1. When the amount of data is large, count query is unnecessary and expensive.
2. Wrapping user's statements with subqueries is also expensive on some database engine, such as mysql 5.6 and earlier version.
  
## 2. Complex SQL query
```
SELECT * FROM test_table2 WHERE id IN (SELECT * FROM test_table LIMIT <offset>, <size>)
```
The following statements will be created automatically by sequel:
* SELECT count(*) AS count FROM ( SELECT * FROM test_table2 WHERE id IN (SELECT * FROM test_table) ) AS t1 LIMIT 1
* SELECT * FROM ( SELECT * FROM test_table2 WHERE id IN (SELECT * FROM test_table) ) AS t1 LIMIT 10 OFFSET 0
* SELECT * FROM ( SELECT * FROM test_table2 WHERE id IN (SELECT * FROM test_table) ) AS t1 LIMIT 10 OFFSET 10
* subsequent ...

But, But what I expect is:
* SELECT * FROM test_table2 WHERE id IN (SELECT * FROM test_table LIMIT 10 OFFSET 0)
* SELECT * FROM test_table2 WHERE id IN (SELECT * FROM test_table LIMIT 10 OFFSET 10)
* subsequent ...

## 3. Stored procedure
```
CALL fetch_my_data(:sql_last_value, <offset>, <size>)
```
The following statements will be created automatically by sequel:
* SELECT count(*) AS count FROM ( CALL fetch_my_data(:sql_last_value) ) AS t1 LIMIT 1
* SELECT * FROM ( CALL fetch_my_data(:sql_last_value) ) AS t1 LIMIT 10 OFFSET 0
* SELECT * FROM ( CALL fetch_my_data(:sql_last_value) ) AS t1 LIMIT 10 OFFSET 10
* subsequent ...

Issues:
1. This syntax is not supported. At least MySQL is.
2. This is not what I want. What I expect is:
    * CALL fetch_my_data(:sql_last_value, 0, 10)
    * CALL fetch_my_data(:sql_last_value, 10, 10)

    I would decide how to page inside the stored procedure.

<br/>

---
Therefore, in order to solve the above problems, as a user, I really need a possibility of writing paging conditions myself.

Now, here's what I can do more:
1. Give `jdbc_paging_avoid_count` a new name that makes it more easier to understand its purpose, such as `jdbc_paging_manual_mode` / `jdbc_paging_custom_mode`. Or you give one.
2. Check the users' statement configuration and force the user to use the `:offset` and `:size` parameters in it.

And WDYT?

<br/>
The following are some related issues in old repository from other users, which also express similar demands, for reference.

https://hub.fastgit.org/logstash-plugins/logstash-input-jdbc/issues/345
https://hub.fastgit.org/logstash-plugins/logstash-input-jdbc/issues/332
https://hub.fastgit.org/logstash-plugins/logstash-input-jdbc/issues/307
https://hub.fastgit.org/logstash-plugins/logstash-input-jdbc/issues/305
https://hub.fastgit.org/logstash-plugins/logstash-input-jdbc/issues/298#issuecomment-416886991
https://hub.fastgit.org/logstash-plugins/logstash-input-jdbc/pull/387/files

Reply:
Thanks @dingx1 for the full explanation, now I understand it better.

I would prefer `jdbc_paging_manual_mode`, the intent seems more explicit.

We need more work on the documentation to expose this functionality more clean as we can, describing the use cases that it fits.

Reply:
Hi, I've renamed it and improved the documentation. Could you please review it again, thanks!